packages/eslint-plugin/src/rules/no-misused-promises.ts

```diff
@@ -130,13 +151,163 @@ export default util.createRule<Options, 'conditional' | 'voidReturn'>({
         const tsNode = parserServices.esTreeNodeToTSNodeMap.get(argument);
         if (returnsThenable(checker, tsNode as ts.Expression)) {
           context.report({
-            messageId: 'voidReturn',
+            messageId: 'voidReturnArgument',
             node: argument,
           });
         }
       }
     }
 
+    function checkAssignment(node: TSESTree.AssignmentExpression): void {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      const varType = checker.getTypeAtLocation(tsNode.left);
+      if (!isVoidReturningFunctionType(checker, tsNode.left, varType)) {
+        return;
+      }
+
+      if (returnsThenable(checker, tsNode.right)) {
+        context.report({
+          messageId: 'voidReturnVariable',
+          node: node.right,
+        });
+      }
+    }
+
+    function checkVariableDeclaration(node: TSESTree.VariableDeclarator): void {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      if (tsNode.initializer === undefined || node.init === null) {
+        return;
+      }
+      const varType = checker.getTypeAtLocation(tsNode.name);
+      if (!isVoidReturningFunctionType(checker, tsNode.initializer, varType)) {
+        return;
+      }
+
+      if (returnsThenable(checker, tsNode.initializer)) {
+        context.report({
+          messageId: 'voidReturnVariable',
+          node: node.init,
+        });
+      }
+    }
+
+    function checkProperty(node: TSESTree.Property): void {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      if (ts.isPropertyAssignment(tsNode)) {
+        const contextualType = checker.getContextualType(tsNode.initializer);
+        if (
+          contextualType !== undefined &&
+          isVoidReturningFunctionType(
+            checker,
+            tsNode.initializer,
+            contextualType,
+          ) &&
+          returnsThenable(checker, tsNode.initializer)
+        ) {
+          context.report({
+            messageId: 'voidReturnProperty',
+            node: node.value,
+          });
+        }
+      } else if (ts.isShorthandPropertyAssignment(tsNode)) {
+        const contextualType = checker.getContextualType(tsNode.name);
+        if (
+          contextualType !== undefined &&
+          isVoidReturningFunctionType(checker, tsNode.name, contextualType) &&
+          returnsThenable(checker, tsNode.name)
+        ) {
+          context.report({
+            messageId: 'voidReturnProperty',
+            node: node.value,
+          });
+        }
+      } else if (ts.isMethodDeclaration(tsNode)) {
+        if (ts.isComputedPropertyName(tsNode.name)) {
+          return;
+        }
+        const obj = tsNode.parent;
+
+        // This condition isn't satisfied but needed for type checking.
+        // The fundamental reason is that mapping between ESLint and TypeScript ASTs is not 1:1.
+        /* istanbul ignore if */ if (!ts.isObjectLiteralExpression(obj)) {
```

Comment:
I think TypeScript and Codecov are is right, actually?

```ts
type O = { f?: () => void };

const obj: O = new (class Foo {
  f?: () => Promise<string>;
});

obj.f?.();
```

Reply:
Code here is invoked when a node matches ESLint's `Property` AST type, and `Property` doesn't include class methods. 

But after converting it to a TypeScript AST node it becomes `MethodDeclaration` which includes both object literal methods and class methods. This is why this `if` is needed :sob:

Indeed, your example would need a check against class methods, but that'd be a tough challenge because we can't use contextual type for this and instead we need to compare two object types. I also worry if that'd slow down this rule because a lot more type comparison would be required to lint out that case.

So maybe that one could be a separate task than this PR? 🥺

Reply:
Gotcha, that makes sense. Could you make that clear in the comment then?

Let's remove the `istanbul ignore if` too. We can ignore a dip in coverage from this particular line.

Reply:
Thanks for understanding, I updated the comment. 🙂